xarray/tests/test_units.py

```diff
@@ -915,24 +958,39 @@ def test_merge_dataarray(variant, unit, error, dtype):
     )
 
     func = function(xr.merge)
-    if error is not None:
+    if error is not None and variant != "data":
```

Comment:
this is necessary because `merge` requires named `DataArray` objects and puts them into a `Dataset`, which means the units don't have to be compatible.